component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -777,42 +780,65 @@ private void processValidUserToken(AuthenticationContext context, AuthenticatedU
     private void checkWithBackUpCodes(AuthenticationContext context, String userToken,
                                       AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
 
-        String savedOTPString = null;
+        String[] savedOTPs = null;
         String username = context.getProperty(SMSOTPConstants.USER_NAME).toString();
         String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
         UserRealm userRealm = getUserRealm(username);
         try {
             if (userRealm != null) {
-                savedOTPString = userRealm.getUserStoreManager()
-                        .getUserClaimValue(tenantAwareUsername, SMSOTPConstants.SAVED_OTP_LIST, null);
+                UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+                if (userStoreManager != null) {
+                    String savedOTPString = userStoreManager
+                            .getUserClaimValue(tenantAwareUsername, SMSOTPConstants.SAVED_OTP_LIST, null);
+                    if (StringUtils.isNotEmpty(savedOTPString)) {
+                        savedOTPs = savedOTPString.split(FrameworkUtils.getMultiAttributeSeparator());
+                    }
+                }
             }
-            if (StringUtils.isEmpty(savedOTPString)) {
+            // Check whether there is any backup OTPs and return.
+            if (ArrayUtils.isEmpty(savedOTPs)) {
                 if (log.isDebugEnabled()) {
                     log.debug("The claim " + SMSOTPConstants.SAVED_OTP_LIST + " does not contain any values");
                 }
                 throw new AuthenticationFailedException("The claim " + SMSOTPConstants.SAVED_OTP_LIST +
                         " does not contain any values");
-            } else if (savedOTPString.contains(userToken)) {
+            }
+            if (isBackUpCodeValid(savedOTPs, userToken)) {
                 if (log.isDebugEnabled()) {
                     log.debug("Found saved backup SMS OTP for user :" + authenticatedUser);
                 }
                 context.setSubject(authenticatedUser);
-                savedOTPString = savedOTPString.replaceAll(userToken, "").replaceAll(",,", ",");
+                savedOTPs = (String[]) ArrayUtils.removeElement(savedOTPs, userToken);
                 userRealm.getUserStoreManager().setUserClaimValue(tenantAwareUsername,
-                        SMSOTPConstants.SAVED_OTP_LIST, savedOTPString, null);
+                        SMSOTPConstants.SAVED_OTP_LIST, String.join(", ", savedOTPs), null);
```

Comment:
Shouldn't we use `FrameworkUtils.getMultiAttributeSeparator()` here also?

Reply:
Not necessarily. This claim is internal to the system. The list of values will always be a comma separated list. And the claim is not meant to be shared with external entities (via JWTs or SAML assertions)

Reply:
This was just a code quality feedback. Doesn't affect security (and maybe not functionality). Up to the team if you want to fix it :)

Reply:
Suggested because you are using that util function in line 794.

Reply:
Previously we have used "," to plit. So if we use MultiAttributeSeparator, it may break in some places. So I will remove it in the line 794 with this commit https://github.com/wso2-extensions/identity-outbound-auth-sms-otp/pull/77/commits/4583b70e27557c1a5d3baab1c082e1e6b66911e4